test/self_protocore/propose_link.js

```diff
@@ -0,0 +1,158 @@
+    config.genesisProposedMetablockHeight = new BN(1);
+
+    config.selfProtocore = await TestSelfProtocore.new();
+    await config.selfProtocore.setCoconsensus(config.coconsensusAddress);
+    await config.selfProtocore.setGenesisStorage(
+      config.genesisParentVoteMessageHash,
+      config.genesisSourceTransitionHash,
+      config.genesisSourceBlockHash,
+      config.genesisSourceBlockNumber,
+      config.genesisTargetBlockHash,
+      config.genesisTargetBlockNumber,
+      config.accumulatedGas,
+    );
+
+    await config.selfProtocore.setup(
+      config.metachainId,
+      config.domainSeparator,
+      config.epochLength,
+      config.metablockHeight,
+    );
+
+    config.sourceKernelHash = Utils.getRandomHash();
+    await config.selfProtocore.setOpenKernelHash(config.sourceKernelHash);
+
+    config.openKernelHeight = new BN(20);
+    await config.selfProtocore.setOpenKernelHeight(config.openKernelHeight);
+  });
+
+  contract('Positive Tests', async () => {
+    it('should propose a valid link', async () => {
+      const targetBlockHash = Utils.getRandomHash();
+      const currentBlockNumber = await web3.eth.getBlockNumber();
+      const targetBlockNumber = config.epochLength.muln(currentBlockNumber);
+
+      const sourceOriginObservation = Utils.getRandomHash();
+      const sourceDynasty = new BN(10);
+      const sourceAccumulatedGas = new BN(10000);
+      const sourceCommitteeLock = Utils.getRandomHash();
+
+      await config.selfProtocore.proposeLink(
+        config.genesisVoteMessageHash,
+        targetBlockHash,
+        targetBlockNumber,
+        config.sourceKernelHash,
+        sourceOriginObservation,
+        sourceDynasty,
+        sourceAccumulatedGas,
+        sourceCommitteeLock,
+      );
+      const sourceTransitionHash = SelfProtocoreUtils.hashSourceTransition(
+        config.domainSeparator,
+        config.sourceKernelHash,
+        sourceOriginObservation,
+        sourceDynasty,
+        sourceAccumulatedGas,
+        sourceCommitteeLock,
+      );
+
+      const voteMessageHash = ProtocoreUtils.hashVoteMessage(
+        config.domainSeparator,
+        sourceTransitionHash,
+        config.genesisTargetBlockHash,
+        targetBlockHash,
+        config.genesisTargetBlockNumber,
+        targetBlockNumber,
+      );
+
+      const voteMessageObject = await config.selfProtocore.links.call(
+        voteMessageHash,
+      );
+
+      assert.strictEqual(
+        voteMessageObject.parentVoteMessageHash,
+        config.genesisVoteMessageHash,
+        'Incorrect parent vote message hash',
+      );
+
+      assert.strictEqual(
+        voteMessageObject.sourceTransitionHash,
+        sourceTransitionHash,
+        'Incorrect source transition hash',
+      );
+
+      assert.strictEqual(
+        voteMessageObject.targetBlockHash,
+        targetBlockHash,
+        'Incorrect target block hash',
+      );
+
+      assert.isOk(targetBlockNumber.eq(voteMessageObject.targetBlockNumber));
```

Comment:
Please add a reason of failure for all the asserts.

Reply:
Done